librz/io/p/io_srec.c

```diff
@@ -0,0 +1,802 @@
+	fclose(out);
+	out = NULL;
+	return 0;
+}
+
+//write contiguous block of data to file; ret 0 if ok
+//max 65535 bytes; 
+static int fwblock(FILE *fd, ut8 *b, ut32 start_addr, ut16 size) {
+	ut8 cks;
+	char linebuf[80];
+	ut16 last_addr;
+	int j;
+	ut32 i; //has to be bigger than size !
+
+	if (size < 1 || !fd || !b) {
+		return -1;
+	}
+
+	for (i = 0; (i + 0x10) < size; i += 0x10) {
+		cks = 0x10;
+		cks += (i + start_addr) >> 8;
+		cks += (i + start_addr);
+		for (j = 0; j < 0x10; j++) {
+			cks += b[j];
+		}
+		cks = 0 - cks;
+		if (fprintf(fd, ":10%04x00%02x%02x%02x%02x%02x%02x%02x"
+				"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x\n",
+			    (i + start_addr) & 0xffff, b[0], b[1], b[2], b[3], b[4], b[5], b[6],
+			    b[7], b[8], b[9], b[10], b[11], b[12], b[13],
+			    b[14], b[15], cks) < 0) {
+			return -1;
+		}
+		start_addr += 0x10;
+		b += 0x10;
+		if ((start_addr & 0xffff) < 0x10) {
+			//addr rollover: write ext address record
+			if (fw04b(fd, start_addr >> 16) < 0) {
+				return -1;
+			}
+		}
+	}
+	if (i == size) {
+		return 0;
+	}
+	//write crumbs
+	last_addr = i + start_addr;
+	cks = -last_addr;
+	cks -= last_addr >> 8;
+	for (j = 0; i < size; i++, j++) {
+		cks -= b[j];
+		sprintf(linebuf + (2 * j), "%02X", b[j]);
+	}
+	cks -= j;
+
+	if (fprintf(fd, ":%02X%04X00%.*s%02X\n", j, last_addr, 2 * j, linebuf, cks) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+static int fw04b(FILE *fd, ut16 eaddr) {
+	ut8 cks = 0 - (6 + (eaddr >> 8) + (eaddr & 0xff));
+	return fprintf(fd, ":02000004%04X%02X\n", eaddr, cks);
+}
+
+static int __read(RzIO *io, RzIODesc *fd, ut8 *buf, int count) {
+	if (!fd || !fd->data || (count <= 0)) {
+		return -1;
+	}
+	Rihex *rih = fd->data;
+	memset(buf, io->Oxff, count);
+	int r = rz_buf_read_at(rih->rbuf, io->off, buf, count);
+	if (r >= 0) {
+		rz_buf_seek(rih->rbuf, r, RZ_BUF_CUR);
+	}
+	return r;
+}
+
+static int __close(RzIODesc *fd) {
+	if (!fd || !fd->data) {
+		return -1;
+	}
+	Rihex *rih = fd->data;
+	rz_buf_free(rih->rbuf);
+	free(rih);
+	fd->data = NULL;
+	return 0;
+}
+
+static ut64 __lseek(struct rz_io_t *io, RzIODesc *fd, ut64 offset, int whence) {
+	Rihex *rih;
+	if (!fd || !fd->data) {
+		return -1;
+	}
+	rih = fd->data;
+	io->off = rz_buf_seek(rih->rbuf, offset, whence);
+	return io->off;
+}
+
+static bool __plugin_open(RzIO *io, const char *pathname, bool many) {
+	return (!strncmp(pathname, SREC_PATH_PREFIX, strlen(SREC_PATH_PREFIX)));
+}
+
+//parsing function
+static bool SREC_parse(RzBuffer *rbuf, char *str) {
+	ut8 *sec_tmp;
+	ut32 sec_start = 0; //addr for next section write
+	ut32 segreg = 0; //basis for addr fields
+	ut32 addr_tmp = 0; //addr for record
+	ut16 next_addr = 0; //for checking if records are sequential
+	char *eol;
+	ut8 cksum;
+	int extH, extL;
+	int bc = 0, byte, i, l;
+	char type; // 1 digit
+	//fugly macro to prevent an overflow of rz_buf_write_at() len
+#define SEC_MAX (sec_size < ST32_MAX) ? sec_size : INT_MAX
+	ut32 sec_size = 0;
+	const int sec_count = UT16_MAX;
```

Comment:
please convert all `int` to `st32`

Reply:
Do you mean all `int` from this function or all `int` from the file

Reply:
any `int` in the file